lib/util/historical-buffer-manager.ts

```diff
@@ -0,0 +1,95 @@
+
+// This class creates a buffer manager which manage N buffers(N is taken from constructor parameter)
+class HistoricalBufferManager<T> {
+
+    // PRIVATE VARIABLES
+    private _buffers: Buffer<T>[];
+    private _currentIndex: number;
+    private _historyDepth: number;
+    private _defaultBuffLength: number;
+
+    // maxHistoryDepth: Maximum numbers of preserved buffer history(default value: 1)
+    // maxBufferLength: Maximum numbers of data item per level of buffer(default value: 100)
+    constructor(maxHistoryDepth: number = 1, maxBufferLength: number = 100) {
+        assert(maxHistoryDepth >= 1, 'Invalid maxHistoryDepth parameter ' + maxHistoryDepth);
+        assert(maxBufferLength >= 1, 'Invalid maxBufferLength parameter ' + maxBufferLength);
+        this._historyDepth = maxHistoryDepth;
+        this._defaultBuffLength = maxBufferLength;
+
+        // Initialize buffers
+        this._buffers = [];
+        this._currentIndex = 0;
+        for (var i = 0; i < this._historyDepth; ++i) {
+            this._buffers.push(new Buffer<T>(this._defaultBuffLength));
+        }
+    }
+
+    // Push one data item to the current buffer
+    pushValue(data: T): void {
+        var currBuffer: Buffer<T> = this._buffers[this._currentIndex];
+        if (currBuffer.data.length === currBuffer.maxBufferLength) {
+            currBuffer.data.splice(0, 1);
+            ++currBuffer.overflow;
+        }
+        currBuffer.data.push(data);
+    }
+
+    // Flush the last data array from the history and move the entire history one step forward
+    // Subsequent push will be stored into a new array
+    // bufferLength: Optional parameter to control the length of the new buffer
+    // Return the buffered data of the most recent flushed array
+    startNewBuffer(bufferLength: number = this._defaultBuffLength): interfaces.IBufferedData<T> {
+        assert(bufferLength >= 1, 'Invalid depth parameter ' + bufferLength);
+        var currBuffer: Buffer<T> = this._buffers[this._currentIndex];
+        var ret: interfaces.IBufferedData<T> = currBuffer.getObject;
+        // Get the new current index
+        this._currentIndex = (this._currentIndex + this._historyDepth - 1) % this._historyDepth;
+        currBuffer = this._buffers[this._currentIndex];
+        currBuffer.maxBufferLength = bufferLength; // Reset max buffer length
+        currBuffer.data.length = 0;  // Clear the current array
+        currBuffer.overflow = 0;  // Reset the overflow counter
+        return ret;
+    }
+
+    // Get buffered data array.
+    // depth: This optional parameter controls which level of history will be retrieved
+    // depth: Default is to retrieve the current buffer. The first history will be depth 1
+    // Return an object containing the data array and an integer for number of overflow items
+    getBuffer(depth: number = 0): interfaces.IBufferedData<T> {
+        assert(depth >= 0 && depth <= this._historyDepth - 1, 'Invalid depth parameter ' + depth);
```

Comment:
Note that there are a couple of places where multiples things are being checked in the same assert besides just this line.